pkg/etcd/init/init.go

```diff
@@ -0,0 +1,126 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright Authors of Cilium
+
+package init
+
+import (
+	"context"
+	"fmt"
+
+	clientv3 "go.etcd.io/etcd/client/v3"
+
+	"github.com/cilium/cilium/pkg/kvstore"
+)
+
+const (
+	rootRoleName = "root"
+)
+
+func InitEtcd(ctx context.Context, client *clientv3.Client, clusterName string) error {
+	_, err := client.Put(ctx, kvstore.HasClusterConfigPath, "true")
+	if err != nil {
+		return err
+	}
+	err = setupRootUser(ctx, client)
+	if err != nil {
+		return err
+	}
+	err = setupAdminUser(ctx, client, clusterName)
+	if err != nil {
+		return err
+	}
+	err = setupExternalWorkloadUser(ctx, client)
+	if err != nil {
+		return err
+	}
+	err = setupRemoteUser(ctx, client)
+	if err != nil {
+		return err
+	}
+	_, err = client.AuthEnable(ctx)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func setupRootUser(ctx context.Context, client *clientv3.Client) error {
+	rootUsername := "root"
+	_, err := client.UserAddWithOptions(ctx, rootUsername, "", &clientv3.UserAddOptions{NoPassword: true})
+	if err != nil {
+		return err
+	}
+	_, err = client.UserGrantRole(ctx, rootUsername, rootRoleName)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func setupAdminUser(ctx context.Context, client *clientv3.Client, clusterName string) error {
```

Comment:
Nit. This function seems a copy of the above one, and could be easily generalized to reduce duplication. The same is likely possible for the ones below (e.g., additionally passing the list of prefixes to be granted).

Reply:
I was hoping to avoid trying to generalise the code honestly. Keeping it written out like this, although repetitive, is clearer to read for those less familiar with the code. One advantage of the shell script before was that it was very clear what it was doing. If you'd still like the generalisation I can try to do that while keeping clarity.

Reply:
It is fine on my side to keep it as now for the other functions. I'd still suggest to merge this one with the root logic, as there seems to be only a difference in the user name.

Reply:
I've restructured the code a lot. What do you think now?

Reply:
:+1: Looks reasonable to me.